pkg/controller/podgc/gc_controller.go

```diff
@@ -124,6 +130,73 @@ func isPodTerminated(pod *v1.Pod) bool {
 	return false
 }
 
+// isPodTerminating returns true if the pod is terminating.
+func isPodTerminating(pod *v1.Pod) bool {
+	if pod.ObjectMeta.DeletionTimestamp != nil {
+		klog.Infof("garbage collecting pod %s that is terminating. Phase [%v]", pod.Name, pod.Status.Phase)
+		return true
+	}
+	return false
+}
+
+// podToleratesTaint returns true if the pod tolerates the provided taint.
+func podToleratesTaint(taint *v1.Taint, pod *v1.Pod) bool {
+	for _, toleration := range pod.Spec.Tolerations {
+		if toleration.ToleratesTaint(taint) {
+			return true
+		}
+	}
+	return false
+}
+
+func (gcc *PodGCController) gcTerminating(pods []*v1.Pod) {
+	klog.V(4).Info("GC'ing terminating pods that are on out of service nodes")
+	terminatingPods := []*v1.Pod{}
+	for _, pod := range pods {
+		if isPodTerminating(pod) {
+			node, err := gcc.nodeLister.Get(string(pod.Spec.NodeName))
+			if err != nil {
+				klog.Errorf("failed to get node:%s", err)
+				continue
+			}
+
+			// Add this pod to terminatingPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `node.kubernetes.io/out-of-service` taint with `NoExecute` effect.
```

Comment:
I thought we had decided the flavor of taint did not matter, but I guess that never got resolved.  https://github.com/kubernetes/enhancements/pull/1116#discussion_r792904049

I think you don't care what taint effect is - you should only care that the pod is terminating and the node is out-of-service.

I don't think (??) you need to check whether the pod tolerates out-of-service or not - you've already confirmed it is terminating.  Is there a reason to do this extra check?

Reply:
agree no need to check taint here. If pod has tolerations, the pod will not be evicted in the first place and is not in terminating

Reply:
>I don't think (??) you need to check whether the pod tolerates out-of-service or not - you've already confirmed it is terminating. Is there a reason to do this extra check?

No reason. I think I made some wrong assumptions. Thank you. 

Have incorporated the fix.